src/main/java/net/fabricmc/mappingio/format/FeatureSetImpl.java

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright (c) 2024 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.mappingio.format;
+
+class FeatureSetImpl implements FeatureSet {
+	FeatureSetImpl(boolean hasNamespaces, MetadataSupport fileMetadata, MetadataSupport elementMetadata, NameSupport packages, NameSupport classes, MemberSupport fields, MemberSupport methods, LocalSupport args, LocalSupport vars, ElementCommentSupport elementComments, boolean hasFileComments) {
+		this.hasNamespaces = hasNamespaces;
+		this.fileMetadata = fileMetadata;
+		this.elementMetadata = elementMetadata;
+		this.packages = packages;
+		this.classes = classes;
+		this.fields = fields;
+		this.methods = methods;
+		this.args = args;
+		this.vars = vars;
+		this.elementComments = elementComments;
+		this.hasFileComments = hasFileComments;
+	}
+
+	@Override
+	public boolean hasNamespaces() {
+		return hasNamespaces;
+	}
+
+	@Override
+	public MetadataSupport fileMetadata() {
+		return fileMetadata;
+	}
+
+	@Override
+	public MetadataSupport elementMetadata() {
+		return elementMetadata;
+	}
+
+	@Override
+	public NameSupport packages() {
+		return packages;
+	}
+
+	@Override
+	public NameSupport classes() {
+		return classes;
+	}
+
+	@Override
+	public MemberSupport fields() {
+		return fields;
+	}
+
+	@Override
+	public MemberSupport methods() {
+		return methods;
+	}
+
+	@Override
+	public LocalSupport args() {
+		return args;
+	}
+
+	@Override
+	public LocalSupport vars() {
+		return vars;
+	}
+
+	@Override
+	public ElementCommentSupport elementComments() {
+		return elementComments;
+	}
+
+	@Override
+	public boolean hasFileComments() {
+		return hasFileComments;
+	}
+
+	private boolean hasNamespaces;
```

Comment:
All these fields should be final.

Reply:
Oops, left-over from when it the feature sets themselves were mutable. Fixed!